File: src/main/java/com/softwareplace/service/ValidatorService.java
package com.softwareplace.service;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

import javax.validation.ConstraintViolation;
import javax.validation.Validator;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import com.softwareplace.exception.IllegalConstraintsException;

@Component
public class ValidatorService {

	private final Validator validator;

	@Autowired
	public ValidatorService(Validator validator) {
		this.validator = validator;
	}

	public <T> void validate(T data) throws IllegalConstraintsException {
		final Set<ConstraintViolation<T>> constraintViolations = validator.validate(data);
		if (!constraintViolations.isEmpty()) {
			throw new IllegalConstraintsException("", getMessagesFromConstraintViolation(constraintViolations));
		}
	}

	private <T> List<String> getMessagesFromConstraintViolation(Set<ConstraintViolation<T>> constraintViolations) {
		final List<String> constraintViolationMessages = new ArrayList<>();
		constraintViolations.forEach(constraintViolation -> {
			final List<String> messages = Arrays.stream(constraintViolation.getMessageTemplate().split(","))
					.collect(Collectors.toList());
			constraintViolationMessages.addAll(messages);
		});
		return constraintViolationMessages;
	}
}

File: src/main/kotlin/com/softwareplace/service/ValidatorService.kt
package com.softwareplace.service

import com.softwareplace.exception.IllegalConstraintsException
import com.softwareplace.extension.addAtStartAsCamelCase
import java.util.*
import java.util.stream.Collectors
import javax.validation.ConstraintViolation
import javax.validation.Validator

open class ValidatorService constructor(
    private val validator: Validator
) {

    @Throws(IllegalConstraintsException::class)
    open fun <T> validate(data: T) {
        val constraintViolations = validator.validate(data)
        if (constraintViolations.isNotEmpty()) {
            throw IllegalConstraintsException("Invalid data input", getMessagesFromConstraintViolation(constraintViolations))
        }
    }

    open fun <T> getMessagesFromConstraintViolation(constraintViolations: Set<ConstraintViolation<T>>): Map<String, MutableList<String>> {
        val errors = HashMap<String, MutableList<String>>()
        constraintViolations.forEach { constraintViolation: ConstraintViolation<T> ->
            val messages = Arrays.stream(constraintViolation.messageTemplate.split(",".toRegex()).toTypedArray())
                .collect(Collectors.toList())

            messages.forEach { message: String ->
                val fieldName: Optional<String> = Arrays.stream(message.split("\\s".toRegex()).toTypedArray()).findFirst()
                fieldName.ifPresent { value: String ->
                    errors.computeIfAbsent(value.addAtStartAsCamelCase("invalid")) { ArrayList() }
                        .add(message)
                }
            }
        }
        return errors
    }
}
